### PR TITLE
Support for adding custom header fields to the output.

### LINKED ISSUE #36: Custom request header fields
Hello,

Would it be possible to add the feature of adding custom request headers to the output?
if this makes sense, are you open to contributions?

### PR DESCRIPTION
**Pull request by arcmmartins**
This allows the injection of custom headers on the output.
For this a new config item was added `http.headers.additional` with the format: `header1:value1,header2:value2`; 

The configured headers will be added to the output. This was done by adding the configured headers at the default builder `DEFAULT_HTTP_REQUEST_BUILDER`.

Unit/integration tests were also implemented.

closes #36

### CODE REVIEW

**Comment by hackaugusto** [comment]
Hi @arcmmartins , thank you for the contribution. We will review the PR and come back to you.

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java [hunk 1]
```diff
@@ -150,6 +154,18 @@ public boolean visible(final String name, final Map<String, Object> parsedConfig
...
+                HTTP_HEADERS_ADDITIONAL,
+                ConfigDef.Type.LIST,
+                Collections.EMPTY_LIST,
+                ConfigDef.Importance.LOW,
+                "Additional headers to forward in the http request in the format header:value separated by a comma",
+                CONNECTION_GROUP,
+                groupCounter++,
+                ConfigDef.Width.MEDIUM,
+                HTTP_HEADERS_ADDITIONAL
+        );
```
Comment (jlprat) [inline src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:167]: I think we should make use of the `Validator` in here and make sure the `http.headers.additional` values are following the right format:

- Comma separated
- Pairs of key value separated with a `:`
- No key is repeated

At the moment if a key is repeated the application will throw an exception only when the additional headers are retrieved from the config (not when the config is loaded). Additionally, if a key-value pair is malformed, this is simply ignored through the `.filter(h -> h.length > 1)` call.

Could you use the `define` variant that takes a `ConfigDef.Validator` checking the previously mentioned points?

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java [hunk 2]
```diff
@@ -434,6 +450,13 @@ public final Long kafkaRetryBackoffMs() {
         return getLong(KAFKA_RETRY_BACKOFF_MS_CONFIG);
     }
 
+    public Map<String, String> getAdditionalHeaders() {
+        return getList(HTTP_HEADERS_ADDITIONAL).stream()
+                .map(s -> s.split(HTTP_HEADERS_ADDITIONAL_DELIMITER))
+                .filter(h -> h.length > 1)
```
Comment (jlprat) [inline src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:456]: if we use the validator approach, then this line won't be needed in here anymore.

File: src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java [hunk 3]
```diff
@@ -319,6 +321,38 @@ void headerContentType() {
         assertEquals("application/json", config.headerContentType());
     }
 
+    @Test
```
Comment (jlprat) [inline src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java:324]: Could you add another test where we successfully check that one can add additional headers that only differ in the case?
For example:
```
"http.headers.additional", "test:value,TEST:valueUpperCase"

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java (hunk 1 above)
Comment (jlprat) [reply src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java:167]: Additionally, could you add a test for this under the `HttpSinkConfigValidationTest` file?

**Review by jlprat — CHANGES_REQUESTED** [review]
Thanks a lot for your PR @arcmmartins! And welcome to our project!
 
The PR looks really good. I found a couple of things that could be improved to make the code more resilient to wrong inputs in the config.

Let us know if something is not clear.

Thanks again for your contribution!

File: src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java (hunk 3 above)
Comment (arcmmartins) [reply src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java:324]: From what i understood header fields are case-insensitive, what would be the use case for allowing "duplicate" case sensitive headers?

Comment (jlprat) [reply src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java:324]: Sorry, my bad. What I mean here is to check for this corner case where we can, indeed, provide headers whose key only differs by their case.
The current implementation already supports it. It's just to make sure this case is covered and any future refactoring takes this into consideration.

For example if we provide these: `"http.headers.additional", "test:value,TEST:valueUpperCase"` we should check that 

```java
assertEquals(2, additionalHeaders.size());
assertEquals("value", additionalHeaders.get("test"));
assertEquals("valueUpperCase", additionalHeaders.get("TEST"));
```

Comment (arcmmartins) [reply src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java:324]: I understood, what I am asking is whether it makes sense for case sensitive field names to be supported or should we also restrict this? (from what I understood header field names are case-insensitive, as specified in [RFC](https://datatracker.ietf.org/doc/html/rfc7230#section-3.2))

Comment (jlprat) [reply src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java:324]: Oh, you are completely right! I missed this part of the specification.
Then we should add this check in the validator, "Keys should be unique in a case insensitive way".

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java (hunk 2 above)
Comment (arcmmartins) [reply src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java]: Removed line

File: src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java (hunk 1 above)
Comment (arcmmartins) [reply src/main/java/io/aiven/kafka/connect/http/config/HttpSinkConfig.java]: Added KeyValuePairListValidator for this and respective tests

File: src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java (hunk 3 above)
Comment (arcmmartins) [reply src/test/java/io/aiven/kafka/connect/http/config/HttpSinkConfigTest.java:324]: Added duplicate keys check in validator that takes into account the header field case

**Review by jlprat — APPROVED** [review]
LGTM!

Thanks @arcmmartins for you contribution!

**Comment by arcmmartins** [comment]
@jlprat It appears I have messed up as I didn't sign my commits, sorry. What would you guys recommend to solve this? Should I just close this PR and reopen a new one (with a comment pointing to this one)?

**Comment by jlprat** [comment]
@arcmmartins No worries, we don't require signed commits from external contributors.
It's all good :)
